Expose GeoPackage to the browser tree

elpaso opened the pull request:
Needs some polishing but it's ready for comments.

# Browser GPKG layers in the browser

Also save connections

![image](https://user-images.githubusercontent.com/142164/28945421-b8b47894-78a6-11e7-9381-39e0ceeb8f05.png)


# TODO

- [x] handle rasters
- [x] handle multi-geometries (now they are just ignored by the filter but they should be supported)
- [x] implement source select dialog for ogr-gpkg and spatialite
- [x] create a new layer action
- [x] replace png icon with svg
- [x] remove action

## TODO that will be addressed in a future PR
- [ ] refactoring as a base class to be used also by spatialite
- [ ] replace spatialite native with spatialite-ogr in the browser
- [ ] more robust check (regexp?) for explicit sublayers in the app addVectorLayer slot
- [x] sync with DB manager #5019 (still needs the signals!)
- [x] automatically add newly created database to the stored connections?
- [ ] drag and drop a shapefile on a gpkg

elpaso committed:
Expose GeoPackage to the browser tree

gioman commented:
Thanks @elpaso ! this fixes a very "annoying" inconsistency in QGIS datasaources support.

nyalldawson commented:
Nice to see this being addressed! I'm wondering what your thoughts are for geopackage (and other compound data sources such as gdb) when directly browsing via folders to the source. It'd be nice to allow these to be expanded out too to see the layers contained within. Would you see this is a totally separate piece of work to what you've done here?

elpaso commented:
@nyalldawson good idea, I need to check wether it is an easy integration or not. But in any event I'm not sure I'll have time to address that within this PR, maybe I'll just focus on adding GPKG and a base class that would also suit for spatialite and leave the filesystem browser for a later PR.

rduivenvoorde commented:
Hi Ale,

Cool, working!

![image](https://user-images.githubusercontent.com/731673/28959469-dbc4f576-78fa-11e7-9372-05c43a26e040.png)

Some thoughts (yeah, I know, you give a finger and they take your arm...).
This also goes for other compound data sources:
- the order the layers are added when doing a multiselect is apparently the alphabetical? In the screenshot I tried to put the rivers on top, by selecting it first... but that did not work: whatever you do to try to create a (selection) order it always is loaded in the order you see it in the browser..
- when selecting it looks like there is one column selected (could maybe be to select the full row)

Some general feature requests:
- a connection I create in the browser, seems to be be used in the DBmanager (plugin)? Would be nice if those connections could be shared... or did I do something wrong?
- as we can add styles to layers in a geopackage, what about a dropdown with available styles to the layers  ... ok going too fast now ;-)

@nyalldawson YES, that would be nice

Another observation (see screenshot), if you enable 'Check file contents' in the Options/Data Sources is that there is a inconsistency in how gpkg files (which contain only one layer) are shown. In the GeoPackage node ALL gpkg files are little barrels, while the browser shows a polygin/point/line symbol for a gpkg which only has one layer (see nl.gpkg vs world.gpkg in screenie below)

![selection_060](https://user-images.githubusercontent.com/731673/28960545-68ecb4bc-78ff-11e7-9188-7f547cfecf8b.png)

But as said: all additional observations, thanks for this work Alessandro! Let's make GeoPackage better!!

Gustry commented:
Nice feature @elpaso !
Also I noticed something, if we have a geopackage with a raster and vectors in it, when we drag and drop this geopackage into the canvas, it doesn't ask for vector layers, it adds straight the raster layer. Can we change this to display vector first (more common I guess) or even better rasters and vectors in the sublayer dialogs?

timlinux commented:
Great to see this PR!

elpaso commented:
@rduivenvoorde thanks for the suggestions, I've added some in the TODO list.

elpaso committed:
Move the default CRS for new layers function to QgsProject

The reason is that it needs to access to project's crs
and it's not only needed by app but also by the providers
(will be)

elpaso committed:
Fixes raster sublayer not apperaring in the browser tree

elpaso committed:
Sip update for qgsproject

elpaso commented on src/providers/gdal/qgsgdaldataitems.cpp:
```diff
@@ -99,7 +99,9 @@ QVector<QgsDataItem *> QgsGdalLayerItem::createChildren()
       childItem = new QgsGdalLayerItem( this, name, mSublayers[i], mSublayers[i] );
       if ( childItem )
-        this->addChildItem( childItem );
+      {
```
@wonder-sk would you please have a look to this commit? I'm not sure why it was calling directly `addChildItem` instead of adding the item to the list of `children`

elpaso committed:
Fix sublayers disappear when refreshing GDAL layers

elpaso committed:
Typo

elpaso committed:
Added a comment to this fertility line :)

elpaso committed:
Small typo

elpaso committed:
Initial support for extension-based scan on multiple layers

elpaso committed:
Added raster support to browser geopackage data items

elpaso committed:
Remove initial : from the sublayers name in the browser tree

elpaso committed:
Changed geopackage icon to svg

elpaso committed:
Fixes two issues with the browser refresh

1. endless loop when a sqlite-based layer is opened, and
   wal+shm files are created. This triggers a loop because
   the directoryChanged signal is emitted again and again ...
   This was the real blocker.

2. when a new files appears in a directory
   the directoryChanged is emitted and OGR/GDAL may fail
   to open the file because the file copy was not yet
   finished.

This commit fixes 1 but the fix for 2 is only a best effort using
a 100 ms timer: I could not find a definitive solution to this issue,
a file could be legitimately opened in streaming mode and it will
always triggers an error, there is apparently no way to get
the QFileSystemWatcher emit a signal when new files are closed
flusehd and not when they are just created.

elpaso committed:
Added file path to debug msg output on errors

elpaso committed:
Added mGeoPackage svg icon

elpaso committed:
Removed PNG version of GPKG icon

elpaso commented on src/core/qgsdataitem.cpp:
```diff
@@ -791,12 +792,22 @@ void QgsDirectoryItem::directoryChanged()
+    // this happens when a new file appears in the directory and
+    // the item's children thread will try to open the file with
+    // GDAL or OGR even if it is still being written.
+    QTimer::singleShot( 100, this, SLOT( refresh() ) );
```
@m-kuhn any better idea than this QTimer hack?

elpaso committed:
Updated file header for geopackage data items

also added "db" to multiple layers extensions

elpaso commented:
@rduivenvoorde I've made some progress and fixed some bugs and glitches unrelated to this PR, but I've not found a decent solution for the inconsistency between the directory browser and the GPKG items: I really don't like to see the two entries (raster and vector) for a single gpkg file as happens in the file browser and that is not an easy fix, so I'd prefer to live with the inconsistency instead of aligning to what is in my opinion the worse display.

elpaso committed:
Delete gpkg connections

elpaso commented:
I'm merging as it is, the remaining issues will be addressed in a future PR, some of them require changes in the ogr/gdal provider (delete layer).
